src/ghci/loaded_module.rs

```diff
@@ -0,0 +1,96 @@
+use std::borrow::Borrow;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::hash::Hasher;
+
+use camino::Utf8Path;
+
+use crate::normal_path::NormalPath;
+
+/// Information about a module loaded into a `ghci` session.
+///
+/// Hashing and equality are determined by the module's path alone.
+#[derive(Debug, Clone, Eq)]
+pub struct LoadedModule {
+    /// The module's source file, like `src/My/Cool/Module.hs`.
+    path: NormalPath,
+
+    /// The module's dotted name, like `My.Cool.Module`.
+    ///
+    /// This is present if and only if the module is loaded by name.
+    ///
+    /// Entries in `:show targets` can be one of two types: module paths or module names (with `.` in
+    /// place of path separators). Due to a `ghci` bug, the module can only be referred to as whichever
+    /// form it was originally added as (see below), so we use this to track how we refer to modules.
+    ///
+    /// See: <https://gitlab.haskell.org/ghc/ghc/-/issues/13254#note_525037>
+    name: Option<String>,
+}
+
+impl LoadedModule {
+    /// Create a new module, loaded by path.
+    pub fn new(path: NormalPath) -> Self {
+        Self { path, name: None }
+    }
+
+    /// Create a new module, loaded by name.
+    pub fn with_name(path: NormalPath, name: String) -> Self {
+        Self {
+            path,
+            name: Some(name),
+        }
+    }
```

Comment:
I think I might have asked this before, but what's the motivation for using these methods instead of using the `LoadedModule` constructor directly?

Reply:
It's the usual encapsulation reason. If you want to be able to write `LoadedModule { path, name: None }` then all of the struct's fields have to be public (and theoretically callers could use this to break invariants).

In reality, `LoadedModule` isn't exported publicly by the library and it would be basically fine to have the fields be public, but this is considered "good practice" so I lean towards it.

(Also constructor names can be a little more informative than setting struct fields directly.)